lib/iris/cube.py

```diff
@@ -1590,15 +1602,34 @@ def cell_methods(self, cell_methods):
         self._cell_methods = tuple(cell_methods) if cell_methods else tuple()
 
     @property
+    def core_data(self):
+        """
+        The data at the core of this cube.
+        May be a numpy array or a dask array.
+
+        """
+        if self._numpy_array is not None:
+            result = self._numpy_array
+        else:
+            result = self._dask_array
+        return result
+
+    @property
     def shape(self):
         """The shape of the data of this cube."""
-        shape = self._my_data.shape
-        return shape
+        return self.core_data.shape
 
     @property
     def dtype(self):
-        """The :class:`numpy.dtype` of the data of this cube."""
-        return self._my_data.dtype
+        if self._dtype is None:
+            result = self.core_data.dtype
+        else:
+            result = self._dtype
+        return result
+
+    @dtype.setter
+    def dtype(self, dtype):
+        self._dtype = dtype
```

Comment:
@marqh :bomb: If the `dtype` changes, then the cubes `metadata` surely also needs to change!